### PR TITLE
Expose theme variables in CSS custom properties

### PR DESCRIPTION
**Pull request by quilicicf**
# What does this PR do?

Exposes all theme variables to CSS custom properties on `:root`.

# Why?

To allow plugin authors to re-use them.

See [linked issue](https://github.com/hakimel/reveal.js/issues/2519).

### CODE REVIEW

**Comment by earboxer** [comment]
I had to update node-sass in my dependent project to 4.13.0 in order to get rid of the `no mixin named css-var` error when trying to compile the styles.

File: css/theme/template/exposer.scss [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  @include css-var(--heading1-text-shadow, $heading1TextShadow);
+  @include css-var(--heading1-size, $heading1Size);
+  @include css-var(--heading2-size, $heading2Size);
+  @include css-var(--heading3-size, $heading3Size);
+  @include css-var(--heading4-size, $heading4Size);
+  @include css-var(--code-font, $codeFont);
+  @include css-var(--link-color, $linkColor);
+  @include css-var(--link-color-hover, $linkColorHover);
+  @include css-var(--selection-background-color, $selectionBackgroundColor);
+  @include css-var(--selection-color, $selectionColor);
```
Comment (earboxer) [inline css/theme/template/exposer.scss]: node-sass is having trouble with this for me:

> no mixin named css-var

Do you want to use

```scss
  --background-color: #{$backgroundColor};
  --main-font: #{$mainFont};
  --main-font-size: #{$mainFontSize};
  --main-color: #{$mainColor};
  --block-margin: #{$blockMargin};
  --heading-margin: #{$headingMargin};
  --heading-font: #{$headingFont};
  --heading-color: #{$headingColor};
  --heading-line-height: #{$headingLineHeight};
  --heading-letter-spacing: #{$headingLetterSpacing};
  --heading-text-transform: #{$headingTextTransform};
  --heading-text-shadow: #{$headingTextShadow};
  --heading-font-weight: #{$headingFontWeight};
  --heading1-text-shadow: #{$heading1TextShadow};
  --heading1-size: #{$heading1Size};
  --heading2-size: #{$heading2Size};
  --heading3-size: #{$heading3Size};
  --heading4-size: #{$heading4Size};
  --code-font: #{$codeFont};
  --link-color: #{$linkColor};
  --link-color-hover: #{$linkColorHover};
  --selection-background-color: #{$selectionBackgroundColor};
  --selection-color: #{$selectionColor};
```

instead?

Comment (quilicicf) [reply css/theme/template/exposer.scss]: I can absolutely do that yes. 
Don't remember how I stumbled upon the syntax in my PR, it was the only thing that worked on another project of mine (and as usual I forgot to document the why :laughing: ). 

I'll get a look at this today if I can.

Comment (quilicicf) [reply css/theme/template/exposer.scss]: Also, it is a superset of another proposed change. I'll link the PR here too so we can close both.

**Comment by quilicicf** [comment]
@earboxer I did a little digging to understand where the hell the css-var came from and I did document it. It was [a hack](https://github.com/sasstools/sass-lint/issues/1161#issuecomment-390537190) to avoid breaking sass-lint. I kinda mis-copied the mixin :man_facepalming: 

Given the SASS is not linted on this repository, it does bring any value to add it but at least if you want to lint in the future you'll have the tip :laughing: 

I have pushed the simpler version (and built it which I forgot last time) and re-wrote history to ~hide my shame~ simplify it.

File: css/theme/template/exposer.scss (hunk 1 above)
Comment (quilicicf) [reply css/theme/template/exposer.scss]: Which you found and closed :+1:

**Comment by quilicicf** [comment]
@earboxer Any idea how to get this merged/released? 
I'd love to have it for [my plugin](https://github.com/quilicicf/reveal-prepare-timing-plugin)

**Comment by earboxer** [comment]
I get the feeling that the volume of MRs is too much for @hakimel.

You can always maintain your own fork (I do that for a project that requires https://github.com/hakimel/reveal.js/pull/2408).

**Comment by quilicicf** [comment]
I figured that much given the number of open PRs yup :-( 
It'd be nice to team up to reduce the truck factor, it seems that currently he is the only one with rights on this repository.
I'm currently writing my own solution but it would benefit the community way more if we could instead give helping hand on this repository ....
Let's see if Hakim finds some time to read this and is willing to get help.

**Comment by earboxer** [comment]
I assume anything that benefits slides.com will be merged in. The benefit to slides.com of merging in this PR is that plugins will have the ability to use theme colors.

If slides.com developed a marketplace for plugins (perhaps an added subscription fee, or pay-per plugin) (possibly with options for revenue sharing), then it would make sense from a business perspective for @hakimel to merge this PR in.

Adding maintainers from the community (who aren't affiliated with slides.com) could potentially lead to changes that break slides.com.

> We contribute back by maintaining reveal.js, an open source HTML presentation framework. 

How does that saying go: "Give a man a fish, feed him for a day, make an open-source project, maintain it for a lifetime." I don't blame the slides.com team for the slowness/lack of review of community PRs.

An alternative is to have a prominent community-minded fork pop up: `reveal.js-ce` (though who wants to maintain it...).

**Comment by quilicicf** [comment]
It makes sense from a business perspective and I'm fine with that but merely answering would be nice IMO. 

I've started an opinionated slide presenter that aims at covering only the features I find relevant and cover the two aspects of my two PRs. 

Prolly won't be called `reveal.js-ce` and be used nearly as widely as Reveal.js but i'ts fun to do and will cover my use case while eventually contributing to the community. 

They can merge my PR or not, depending on what makes sense for them from a business standpoint, at least it's already there for the picking.

**Comment by hakimel** [comment]
This is great, thanks @quilicicf!

I've had very little time for reveal.js in the past few years—having two kids, renovating a house and starting a small business pushed OSS work way down on the priority list. Trying to make a big push towards reveal.js 4.0 at the moment (see the dev branch) which I'm excited about!

**Comment by quilicicf** [comment]
It's nice to see Reveal getting attention again, thanks @hakimel !
I can contribute something to the documentation to hint plugin developers about custom CSS variables so they can re-use it. 
Just tell me where you think it would make sense to document this and I'll propose a PR.

**Comment by hakimel** [comment]
That'd be great, thanks! There's a Theming section in the readme that would be a good fit.

**Comment by quilicicf** [comment]
OK I've setup a reminder to do that next time I get a few spare minutes :+1:

**Comment by quilicicf** [comment]
I must be tired, I forgot I already added a line in the `theming` chapter about exposed CSS variables :sweat_smile: 

I think it tells enough as is.

**Comment by hakimel** [comment]
Really? Well that does it then! Thanks :D

**Pull request merged**